### PR TITLE
[PLAYER-4648] Fixed UI lag when user begin seeking for tvOS

### PR DESCRIPTION
**Pull request by MarkEPAM**
Added immediately progress bar updates for seeking process.

Ticket: https://jira.corp.ooyala.com/browse/PLAYER-4648

### CODE REVIEW

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 1]
```diff
@@ -258,235 +263,253 @@ - (void)hideProgressBar {
...
+  self.tableList = [NSMutableArray array];
+  NSMutableArray *subtitleArray = [NSMutableArray array];
+  NSArray *closedCaptionsArray;
+
+  // For each language in the list, add it to the necessary array
+  for (NSString *language in providedList) {
+
+    // If this was the closed captions language, put 'None' and 'CC' langauges in the array
+    if ([language compare: @"cc"] == NSOrderedSame) {
+      closedCaptionsArray = [NSArray arrayWithObjects:[[OOOoyalaTVPlayerViewController currentLanguageSettings] objectForKey:@"Off"], [[OOOoyalaTVPlayerViewController currentLanguageSettings] objectForKey:@"Use Closed Captions"], nil];
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: replace `objectForKey` with shorthand `[]` syntax

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 2]
```diff
@@ -258,235 +263,253 @@ - (void)hideProgressBar {
...
+    } else {
+      if (subtitleArray.count == 0) {
+        [subtitleArray addObject:[[OOOoyalaTVPlayerViewController currentLanguageSettings] objectForKey:@"Off"]];
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: same here

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 3]
```diff
@@ -176,49 +179,51 @@ - (void)setupViewController {
...
+  
+  [NSNotificationCenter.defaultCenter addObserver:self
+                                         selector:@selector(timeChangedNotification)
+                                             name:OOOoyalaPlayerTimeChangedNotification
+                                           object:self.player];
 }
 
 - (void)removeObservers {
-  [[NSNotificationCenter defaultCenter] removeObserver:self];
+  [NSNotificationCenter.defaultCenter removeObserver:self];
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m:194]: is it needed?

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 4]
```diff
@@ -258,235 +263,253 @@ - (void)hideProgressBar {
...
-    NSDictionary *ja = [NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:@"ライブ", @"完了", @"言語", @"さらに詳しく", @"このアプリからビデオをキャストできます", @"切断", @"デバイスに接続",@"Subtitles",@"Off",@"Use Closed Captions", nil] forKeys:keys];
-    NSDictionary *es = [NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:@"En vivo", @"Hecho", @"Idioma", @"Más información", @"Listo para trasmitir videos desde esta aplicación", @"Desconectar", @"Conectar al dispositivo",@"Subtítulos",@"Off",@"Usar Subtítulos", nil] forKeys:keys];
-    
-    OOOoyalaPlayerViewControllerAvailableLocalizations = [NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:en, ja, es, nil] forKeys:[NSArray arrayWithObjects:@"en", @"ja", @"es", nil]];
++ (void)loadDefaultLocale {
+  //LOG(@"Default Localizations Loaded");
+  NSArray *keys = [NSArray arrayWithObjects:@"LIVE", @"Done", @"Languages", @"Learn More", @"Ready to cast videos from this app", @"Disconnect", @"Connect To Device",@"Subtitles",@"Off",@"Use Closed Captions", nil];
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: It should be created with shorthand syntax, i.e. `@[...]`

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 5]
```diff
@@ -258,235 +263,253 @@ - (void)hideProgressBar {
...
+  NSDictionary *en = [NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:@"LIVE", @"Done", @"Languages", @"Learn More", @"Ready to cast videos from this app", @"Disconnect", @"Connect To Device",@"Subtitles",@"Off",@"Use Closed Captions", nil] forKeys:keys];
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: It should be created with shorthand syntax, i.e. `@{...}`

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 6]
```diff
@@ -258,235 +263,253 @@ - (void)hideProgressBar {
...
+  NSDictionary *ja = [NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:@"ライブ", @"完了", @"言語", @"さらに詳しく", @"このアプリからビデオをキャストできます", @"切断", @"デバイスに接続",@"Subtitles",@"Off",@"Use Closed Captions", nil] forKeys:keys];
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: Same here

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 7]
```diff
@@ -258,235 +263,253 @@ - (void)hideProgressBar {
...
+  NSDictionary *es = [NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:@"En vivo", @"Hecho", @"Idioma", @"Más información", @"Listo para trasmitir videos desde esta aplicación", @"Desconectar", @"Conectar al dispositivo",@"Subtítulos",@"Off",@"Usar Subtítulos", nil] forKeys:keys];
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: Same here

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 8]
```diff
@@ -258,235 +263,253 @@ - (void)hideProgressBar {
...
+
+  OOOoyalaPlayerViewControllerAvailableLocalizations = [NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:en, ja, es, nil] forKeys:[NSArray arrayWithObjects:@"en", @"ja", @"es", nil]];
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: Same here

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 9]
```diff
@@ -258,235 +263,253 @@ - (void)hideProgressBar {
...
-        [self useLanguageStrings:[OOOoyalaPlayerViewControllerAvailableLocalizations objectForKey:language]];
-    } else {
-        [self chooseBackupLanguage:language];
-    }
+  if (!OOOoyalaPlayerViewControllerAvailableLocalizations) {
+    [self loadDefaultLocale];
+  }
+  NSString *language = [NSLocale preferredLanguages].firstObject;
+  if ([OOOoyalaPlayerViewControllerAvailableLocalizations objectForKey:language]) {
+    [self useLanguageStrings:[OOOoyalaPlayerViewControllerAvailableLocalizations objectForKey:language]];
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: `objectForKey` -> `[]`

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 10]
```diff
@@ -258,235 +263,253 @@ - (void)hideProgressBar {
...
-+ (NSDictionary*)getLanguageSettings:(NSString *)language {
-    if(!OOOoyalaPlayerViewControllerAvailableLocalizations) {
-        [self loadDefaultLocale];
-    }
-    return [OOOoyalaPlayerViewControllerAvailableLocalizations objectForKey:language];
++ (NSDictionary *)getLanguageSettings:(NSString *)language {
+  if (!OOOoyalaPlayerViewControllerAvailableLocalizations) {
+    [self loadDefaultLocale];
+  }
+  return [OOOoyalaPlayerViewControllerAvailableLocalizations objectForKey:language];
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: `objectForKey` -> `[]`

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 11]
```diff
@@ -258,235 +263,253 @@ - (void)hideProgressBar {
...
+  // Make the pairs <section name, language/presentation list> to save in the langauge array
+  if (closedCaptionsArray) {
+    Pair *closedCaptionsPair = [Pair alloc];
+    closedCaptionsPair.name = @"Closed Captions";
+    closedCaptionsPair.value = closedCaptionsArray;
+    [self.tableList addObject:closedCaptionsPair];
+  }
+  if (subtitleArray.count > 0) {
+    Pair *subtitlePair = [Pair alloc];
+    subtitlePair.name = [[OOOoyalaTVPlayerViewController currentLanguageSettings] objectForKey:@"Languages"];
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: `objectForKey` -> `[]`

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 12]
```diff
@@ -258,235 +263,253 @@ - (void)hideProgressBar {
...
-            videoRect.size.height = videoRect.origin.y + videoRect.size.height - bottomControlsRect.size.height;
-        }
+  CGRect videoRect = [self.player videoRect];
+  if (!hidden) {
+    if (bottomControlsRect.origin.y < videoRect.origin.y + videoRect.size.height) {
+      videoRect.size.height = videoRect.origin.y + videoRect.size.height - bottomControlsRect.size.height;
     }
-    [self.closedCaptionsView setFrame:videoRect];
+  }
+  [self.closedCaptionsView setFrame:videoRect];
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: I guess it's better to write with dot notation. i.e.
```suggestion
  self.closedCaptionsView.frame = videoRect;
```

**Review by maxkupetskii — CHANGES_REQUESTED** [review]

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m (hunk 3 above)
Comment (MarkEPAM) [reply sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m:194]: sometimes we need to remove self observer without dealloc

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m (hunk 1 above)
Comment (MarkEPAM) [reply sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: done

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m (hunk 2 above)
Comment (MarkEPAM) [reply sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: done

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m (hunk 5 above)
Comment (MarkEPAM) [reply sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: done

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m (hunk 8 above)
Comment (MarkEPAM) [reply sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: done

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m (hunk 9 above)
Comment (MarkEPAM) [reply sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: done

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m (hunk 10 above)
Comment (MarkEPAM) [reply sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: done

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m (hunk 11 above)
Comment (MarkEPAM) [reply sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: done

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m (hunk 12 above)
Comment (MarkEPAM) [reply sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: done

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 13]
```diff
@@ -258,235 +263,254 @@ - (void)hideProgressBar {
...
-        [OOOoyalaTVPlayerViewController chooseBackupLanguage:language];
-    }
-    
-    [self refreshClosedCaptionsView];
+- (void)changeLanguage:(NSString *)language {
+  if (!OOOoyalaPlayerViewControllerAvailableLocalizations) {
+    [OOOoyalaTVPlayerViewController loadDefaultLocale];
+  }
+
+  if (language == nil) {
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]:
```suggestion
  if (!language) {
```

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 14]
```diff
@@ -258,235 +263,254 @@ - (void)hideProgressBar {
...
++ (void)loadDefaultLocale {
+  //LOG(@"Default Localizations Loaded");
+  NSArray *keys = @[@"LIVE", @"Done", @"Languages", @"Learn More", @"Ready to cast videos from this app", @"Disconnect", @"Connect To Device",@"Subtitles",@"Off",@"Use Closed Captions"];
+  NSDictionary *en = [NSDictionary dictionaryWithObjects:@[@"LIVE", @"Done", @"Languages", @"Learn More", @"Ready to cast videos from this app", @"Disconnect", @"Connect To Device",@"Subtitles",@"Off",@"Use Closed Captions"] forKeys:keys];
+  NSDictionary *ja = [NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:@"ライブ", @"完了", @"言語", @"さらに詳しく", @"このアプリからビデオをキャストできます", @"切断", @"デバイスに接続",@"Subtitles",@"Off",@"Use Closed Captions", nil] forKeys:keys];
+  NSDictionary *es = [NSDictionary dictionaryWithObjects:@[@"En vivo", @"Hecho", @"Idioma", @"Más información", @"Listo para trasmitir videos desde esta aplicación", @"Desconectar", @"Conectar al dispositivo",@"Subtítulos",@"Off",@"Usar Subtítulos"] forKeys:keys];
+  
+  OOOoyalaPlayerViewControllerAvailableLocalizations = [NSDictionary dictionaryWithObjects:@[en, ja, es] forKeys:@[@"en", @"ja", @"es"]];
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]:
```suggestion
  OOOoyalaPlayerViewControllerAvailableLocalizations = @{@"en": en, @"ja": ja, @"es": es}; 
```

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m [hunk 15]
```diff
@@ -258,235 +263,254 @@ - (void)hideProgressBar {
...
-    
-    OOOoyalaPlayerViewControllerAvailableLocalizations = [NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:en, ja, es, nil] forKeys:[NSArray arrayWithObjects:@"en", @"ja", @"es", nil]];
++ (void)loadDefaultLocale {
+  //LOG(@"Default Localizations Loaded");
+  NSArray *keys = @[@"LIVE", @"Done", @"Languages", @"Learn More", @"Ready to cast videos from this app", @"Disconnect", @"Connect To Device",@"Subtitles",@"Off",@"Use Closed Captions"];
+  NSDictionary *en = [NSDictionary dictionaryWithObjects:@[@"LIVE", @"Done", @"Languages", @"Learn More", @"Ready to cast videos from this app", @"Disconnect", @"Connect To Device",@"Subtitles",@"Off",@"Use Closed Captions"] forKeys:keys];
+  NSDictionary *ja = [NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:@"ライブ", @"完了", @"言語", @"さらに詳しく", @"このアプリからビデオをキャストできます", @"切断", @"デバイスに接続",@"Subtitles",@"Off",@"Use Closed Captions", nil] forKeys:keys];
```
Comment (maxkupetskii) [inline sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: Shorthand syntax for `array` here

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m (hunk 9 above)
Comment (maxkupetskii) [reply sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: I can't see it :(

**Review by maxkupetskii — CHANGES_REQUESTED** [review]

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m (hunk 15 above)
Comment (MarkEPAM) [reply sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: done

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m (hunk 14 above)
Comment (MarkEPAM) [reply sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: done

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m (hunk 13 above)
Comment (MarkEPAM) [reply sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]: done

**Review by maxkupetskii — APPROVED** [review]

File: sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m (hunk 9 above)
Comment (maxkupetskii) [reply sdk/tvOS/OoyalaTVSkinSDK/OOOoyalaTVPlayerViewController.m]:
```suggestion
    [self useLanguageStrings:OOOoyalaPlayerViewControllerAvailableLocalizations[language]];
```

**Review by maxkupetskii — CHANGES_REQUESTED** [review]

**Pull request merged**